src/lib/crypto/kimchi_backend/gadgets/generic.ml

```diff
@@ -0,0 +1,78 @@
+open Core_kernel
+
+open Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint
+
+(* Generic and gate gadget *)
+let add (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (left_input : Circuit.Field.t) (right_input : Circuit.Field.t) :
+    Circuit.Field.t =
+  let open Circuit in
+  (* Witness computation; sum = left_input + right_input *)
+  let sum =
+    exists Field.typ ~compute:(fun () ->
+        let left_input = As_prover.read Field.typ left_input in
+        let right_input = As_prover.read Field.typ right_input in
+        Field.Constant.add left_input right_input )
+  in
+
+  (* Set up generic add gate *)
+  with_label "generic_add_gadget" (fun () ->
+      assert_
+        { annotation = Some __LOC__
+        ; basic =
+            Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint.T
+              (Basic
+                 { l = (Field.Constant.one, left_input)
+                 ; r = (Field.Constant.one, right_input)
+                 ; o = (Option.value_exn Field.(to_constant (negate one)), sum)
+                 ; m = Field.Constant.zero
+                 ; c = Field.Constant.zero
+                 } )
+        } ;
+      sum )
+
+let%test_unit "generic add gadget" =
+  (* Import the gadget test runner *)
+  let open Kimchi_gadgets_test_runner in
+  (* Initialize the SRS cache. *)
+  let () = Kimchi_pasta.Vesta_based_plonk.Keypair.set_urs_info [] in
+
+  (* Helper to test generic add gate gadget
+   *   Inputs operands and expected output: left_input + right_input = sum
+   *   Returns true if constraints are satisfied, false otherwise.
+   *)
+  let test_generic_add left_input right_input sum =
+    try
+      let _proof_keypair, _proof =
+        Runner.generate_and_verify_proof (fun () ->
+            let open Runner.Impl in
+            (* Set up snary variables for inputs and outputs *)
+            let left_input =
+              exists Field.typ ~compute:(fun () ->
+                  Field.Constant.of_int left_input )
+            in
+            let right_input =
+              exists Field.typ ~compute:(fun () ->
+                  Field.Constant.of_int right_input )
+            in
+            let sum =
+              exists Field.typ ~compute:(fun () -> Field.Constant.of_int sum)
+            in
+            (* Use the generic and gate gadget *)
+            let result = add (module Runner.Impl) left_input right_input in
+            Field.Assert.equal sum result ;
+            (* Pad with a "dummy" constraint b/c Kimchi requires at least 2 *)
+            Boolean.Assert.is_true (Field.equal sum sum) )
+      in
+      true
+    with _ -> false
+  in
+
+  (* Positive tests *)
+  assert (Bool.equal (test_generic_add 0 0 0) true) ;
+  assert (Bool.equal (test_generic_add 1 2 3) true) ;
+  (* Negatve tests *)
+  assert (Bool.equal (test_generic_add 1 0 0) false) ;
+  assert (Bool.equal (test_generic_add 2 4 7) false) ;
```

Comment:
I would tend to have 4 different `let%test_unit` for these 4 tests, for example within a `let%test_module`. Do you think it is worthwhile to do so?

Reply:
See https://github.com/MinaProtocol/mina/blob/ccd92e1773344b069695b36d095135b2e1e271dd/src/lib/pickles/wrap.ml#L293

Reply:
I'm up for learning new things :-)

Reply:
If we adopt this approach, does it need to be invoked with the in-line test runner?

```
   dune exec src/lib/pickles/.pickles.inline-tests/inline_test_runner_pickles.exe \
   --profile=dev --display short -- inline-test-runner pickles -only-test wrap.ml
```
?

Reply:
if so, that could be a problem.  Perhaps this is best for a subsequent PR?

Reply:
Not necessarily. It works with `runtest` as well.

Reply:
Create https://github.com/MinaProtocol/mina/issues/12751